### PR TITLE
WIP: Experimental load-shedding for Report calls to Mixer

### PR DESCRIPTION
**Pull request by douglas-reid**
Exploration of one aspect of the approaches discussed in various P&T meetings around the need for load-shedding within the Mixer, particularly as it applies to handling of `Report` calls.

Testing:

```
$ KUBECONFIG=~/.kube/config ~/go/out/linux_amd64/release/mixs server --configStoreURL=fs://$(pwd)/mixer/testdata/config --apiWorkerPoolSize=1
```

```
$ mixc report -c 2 -r 4 -s context.protocol="http",destination.service="test",request.path="/doug-test",destination.uid="kubernetes://productpage-v1-79b8db5f94-dg4pw.crazy" -t request.time="2018-01-29T12:12:14Z",response.time="2018-01-29T12:12:15Z" --bytes_attributes destination.ip="a:3c:00:1a" --stringmap_attributes request.headers=x-b3-traceid:2141234123412344abfe 
...
READY
Report RPC returned Unavailable (Work queue is full (1 pending requests). Try again later.)
Report RPC returned Unavailable (Work queue is full (1 pending requests). Try again later.)
Report RPC returned OK
Report RPC returned Unavailable (Work queue is full (1 pending requests). Try again later.)
Report RPC returned OK
Report RPC returned Unavailable (Work queue is full (1 pending requests). Try again later.)
Report RPC returned Unavailable (Work queue is full (1 pending requests). Try again later.)
Report RPC returned OK

```

This is meant to provoke discussion and further exploration of the space. As such, this PR should NOT be considered for merging at this time.

/hold

### CODE REVIEW

File: mixer/pkg/api/grpcServer.go [hunk 1]
```diff
@@ -216,16 +221,24 @@ var reportResp = &mixerpb.ReportResponse{}
 
 // Report is the entry point for the external Report method
 func (s *grpcServer) Report(ctx context.Context, req *mixerpb.ReportRequest) (*mixerpb.ReportResponse, error) {
+	qrs := atomic.LoadInt64(&s.queuedRequests)
+	if qrs > s.queueLength {
```
Comment (geeknoid) [inline mixer/pkg/api/grpcServer.go]: I think that's too strict, it doesn't allow for burstiness. How about something like queueLength * 8?

**Comment by geeknoid** [comment]
I think as the Mixer has evolved over time, we've ended up in a slightly silly place regarding goroutines.

- At the top, we have gRPC which is polling the port. Whenever a message arrives, it cranks up a new goroutine and calls grpcServer.go with it.

- grrpcServer.go does its work and eventually calls the dispatchToHandler function in pkg/runtime/dispatcher.

- dispatchToHandler invokes N goroutines out of our pool which do the work of calling out to the handler logic. If there aren't enough goroutines in our pool, dispatchToHandler blocks until some goroutines are available.

- Once dispatchToHandler has scheduled all the work, it blocks waiting for all the goroutines it started to complete.

Seems like we're effectively wasting the goroutine that gRPC started for us. The goroutine is there, we might as well use it and avoid the context switch overhead. This would cover most use cases. Our pool of goroutines ends up just being used in cases where we dispatch stuff to multiple handlers at the same time.

Ignoring the above for a moment, I think the right thing to measure given the current code are the number of goroutines that are blocked in dispatchToHandler. Is this number is > <magic constant> then we can consider Mixer too busy and start rejecting requests. The number can be configured on the command-line, with a default of perhaps 1024?

So we count the number of goroutines blocked in dispatchToHandler, and check right at the top of Check and Report by calling a new interface function on the dispatcher interface.

WDYT?

File: mixer/pkg/api/grpcServer.go (hunk 1 above)
Comment (douglas-reid) [reply mixer/pkg/api/grpcServer.go]: Agreed. This was more PoC than anything else. Wanted to have *some* rubric that was easy to test initially for discussion.

**Comment by douglas-reid** [comment]
I initially started with adding a new method to `Dispatcher`, so I'm fine with such an approach in abstract. I think we probably want to support different policies for `Check` and `Report`, though maybe that becomes more of a client-side concern.

Should we consider just using a (configurable) rate-limiter (instead of trying to measure blocked items)?

I agree that the ratio (or criteria) should be command-line controllable.

File: mixer/pkg/api/grpcServer.go (hunk 1 above)
Comment (mandarjog) [reply mixer/pkg/api/grpcServer.go]: Reports arrive in large batches. At this point in the code, we want to make a judgement about whether we can service 1000 new reports in a reasonable amount of time.  Treating all requests equally is not correct. We could assume that all requests cost the same and do it here, but the rejection should be based on response time violations.

Comment (douglas-reid) [reply mixer/pkg/api/grpcServer.go]: @mandarjog are you suggesting that we shed current load based on prior response times?

**Comment by douglas-reid** [comment]
Updated to reflect alternate option of using a `rate.Limiter` for coarse-grained protection of workload.

**Comment by douglas-reid** [comment]
Added a third commit showing what I think is what you were suggesting @geeknoid . I couldn't think of a good way to count blocked goroutines, so I added a secondary overflow queue (same size as original queue in this example).

I'm not sure which option is simpler and easier to track, however.

I think we *definitely* want to add client-side bits as well (including timeouts that carry forward to cancellations, etc., in the Mixer runtime).

I'll start working on a doc for short-term approaches (that doesn't attempt to tackle priority, etc.)

**Comment by douglas-reid** [comment]
Closing in favor of #9146

**Pull request closed without merging**